hibernate-core/src/test/java/org/hibernate/orm/test/entitygraph/FetchGenericAttributesTest.java

```diff
@@ -0,0 +1,253 @@
+
+@DomainModel(
+		annotatedClasses = {
+				FetchGenericAttributesTest.AbsOne.class,
+				FetchGenericAttributesTest.AbsTwo.class,
+				FetchGenericAttributesTest.One.class,
+				FetchGenericAttributesTest.Two.class,
+				FetchGenericAttributesTest.Three.class
+		}
+)
+@SessionFactory
+@JiraKey("HHH-18151")
+public class FetchGenericAttributesTest {
+
+	private Long oneId;
+	private Long threeId;
+
+	@BeforeEach
+	void before(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					final var one = new One();
+					final var two = new Two();
+					final var three = new Three();
+					one.setTwo( two );
+					two.getOnes().add( one );
+					two.setThree( three );
+					three.getTwos().add( two );
+					one.setOneProp( "oneProp" );
+					two.setTwoProp( "twoProp" );
+					three.setThreeProp( "threeProp" );
+
+					session.persist( one );
+					session.persist( two );
+					session.persist( three );
+					oneId = one.getId();
+					threeId = three.getId();
+				}
+		);
+	}
+
+	@AfterEach
+	void after(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					session.createMutationQuery( session.getCriteriaBuilder().createCriteriaDelete( One.class ) )
+							.executeUpdate();
+					session.createMutationQuery( session.getCriteriaBuilder().createCriteriaDelete( Two.class ) )
+							.executeUpdate();
+					session.createMutationQuery( session.getCriteriaBuilder().createCriteriaDelete( Three.class ) )
+							.executeUpdate();
+				}
+		);
+	}
+
+	@Test
+	@JiraKey("HHH-18151")
+	void HHH18151WithGraphParser(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					One one = session.byId( One.class )
+							.with( GraphParser.parse( One.class, "two(three)", session ), GraphSemantic.FETCH )
+							.load( oneId );
+
+					assertThat( one.getId(), is( oneId ) );
+					assertThat( one.getTwo(), isInitialized() );
+					assertThat( one.getTwo().getThree(), isInitialized() );
+				}
+		);
+	}
+
+	@Test
+	@JiraKey("HHH-18151")
+	void HHH18151WithEntityGraph(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					final var oneGraph = session.createEntityGraph( One.class );
+
+					//before the fix you had to use method
+					//oneGraph.addSubgraph( "two", Two.class )
+					//now it works without it.
+					//Maybe we could deprecate methods like this one..._
+					//org.hibernate.graph.spi.GraphImplementor.addSubGraph(java.lang.String, java.lang.Class<AJ>)
+
+					final var twoSubgraph = oneGraph.addSubgraph( "two" );
+
+					twoSubgraph.addAttributeNode( "three" );
+					One one = session.byId( One.class )
+							.with( oneGraph, GraphSemantic.FETCH )
+							.load( oneId );
+
+					assertThat( one.getId(), is( oneId ) );
+					assertThat( one.getTwo(), isInitialized() );
+					assertThat( one.getTwo().getThree(), isInitialized() );
+				}
+		);
+	}
+
+	@Entity
+	public static class One extends AbsOne<Two> {
```

Comment:
It's generally a good idea to use explicit entity names for inner classes to avoid long table names (unsupported by some old db versions we only test on the nightly CI). Please also do that for the other entities in the test.

```suggestion
	@Entity(name = "One")
	public static class One extends AbsOne<Two> {
```